### PR TITLE
Verify Block Proposer RANDAO

### LINKED ISSUE #665: Implement RANDAO multi-hashing
Let's open a PR to address RANDAO multi-hashing. As described by Vitalik [here](https://github.com/ethereum/eth2.0-specs/pull/69), the nodes on the beacon chain will require more hashes for a RANDAO reveal as more time passes.
This is to reduce RANDAO reveal happens during an orphaned block to be used for future blocks.

### PR DESCRIPTION
**Pull request by rauljordan**
This builds off #1083 and resolves #665.

---

# Description

As a follow-up to #1073, during each state transition in which we receive a block, we need to:

- Verify proposer's signature (Pending BLS)
- **Verify block proposer's RANDAO and update the state's latest randao mixes**
- Append candidate pow chain roots to the state

This PR implements verification and processing a block proposer's RANDAO reveal using outlined above using the official spec's RANDAO processing section at commit hash [9469ca78c4269f901029f13e198337419ffb36d5](https://github.com/ethereum/eth2.0-specs/commit/9469ca78c4269f901029f13e198337419ffb36d5). This PR also deprecates the state's `randao_mix` field.

### CODE REVIEW

**Comment by rauljordan** [comment]
This is pending changes from https://github.com/ethereum/eth2.0-specs/pull/327/files

**Comment by rauljordan** [comment]
@terenc3t we don't have to clear it, we just have to update it according to
```
Set state.latest_randao_mixes[state.slot % LATEST_RANDAO_MIXES_LENGTH] = xor(state.latest_randao_mixes[state.slot % LATEST_RANDAO_MIXES_LENGTH], block.randao_reveal)
```

File: beacon-chain/core/blocks/block_operations.go [hunk 1]
```diff
@@ -44,6 +44,62 @@ func ProcessPOWReceiptRoots(
...
+		latestMix[i] ^= x
+	}
+	proposer.RandaoCommitmentHash32 = block.GetRandaoRevealHash32()
+	proposer.RandaoLayers = 0
+	registry[proposerIndex] = proposer
+	beaconState.LatestRandaoMixesHash32S[beaconState.GetSlot()%latestMixesLength] = latestMix[:]
+	beaconState.ValidatorRegistry = registry
+	return beaconState, nil
+}
+
```
Comment (nisdas) [inline beacon-chain/core/blocks/block_operations.go]: Can you add the pseudocode from the spec here ?

File: beacon-chain/core/randao/randao.go [hunk 2]
```diff
@@ -21,3 +22,13 @@ func UpdateRandaoLayers(state *pb.BeaconState, slot uint64) (*pb.BeaconState, er
 	state.ValidatorRegistry = vreg
 	return newState, nil
 }
+
+// UpdateRandaoMixes sets the beacon state's latest randao mixes according to the latest
+// beacon slot.
+func UpdateRandaoMixes(state *pb.BeaconState) *pb.BeaconState {
+	newState := proto.Clone(state).(*pb.BeaconState)
```
Comment (nisdas) [inline beacon-chain/core/randao/randao.go]: remove cloning of state

File: beacon-chain/blockchain/service.go [hunk 3]
```diff
@@ -224,34 +239,24 @@ func (c *ChainService) blockProcessing(processedBlock chan<- *pb.BeaconBlock) {
 		// can be received either from the sync service, the RPC service,
 		// or via p2p.
 		case block := <-c.incomingBlockChan:
-
```
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply beacon-chain/blockchain/service.go]: +1 why delete this?

**Comment by rauljordan** [comment]
PTAL @nisdas

File: beacon-chain/blockchain/service.go [hunk 4]
```diff
@@ -77,25 +77,6 @@ func NewChainService(ctx context.Context, cfg *Config) (*ChainService, error) {
 // Start a blockchain service's main event loop.
 func (c *ChainService) Start() {
 	log.Info("Starting service")
-
-	var err error
```
Comment (nisdas) [inline beacon-chain/blockchain/service.go]: This whole chunk shouldn't be deleted, we need it for block pre-processing : 
```go
	var err error
	c.genesisTime, err = c.beaconDB.GenesisTime()
	if err != nil {
		log.Fatalf("Unable to retrieve genesis time, therefore blockchain service cannot be started %v", err)
		return
	}
```

File: beacon-chain/core/blocks/block_operations.go (hunk 1 above)
Comment (nisdas) [reply beacon-chain/core/blocks/block_operations.go]: add this part here in a comment, just so we know whats happening from the spec 
```
 Verify that repeat_hash(block.randao_reveal, proposer.randao_layers) == proposer.randao_commitment.
```

**Pull request merged**